api/controllers/Irc_EntryController.js

```diff
@@ -117,21 +117,69 @@ module.exports = {
     switch (request_body.operation) {
     case Event.OPERATION_INTER_SITE_TRANSFER:
       return processEventDetainee(request_body)
-          .then((detainee) => this.handleInterSiteTransfer(detainee, request_body));
+        .then((detainee) => this.handleInterSiteTransfer(detainee, request_body));
     case Event.OPERATION_UPDATE:
       return processEventDetainee(request_body)
-          .tap((detainee) => Centres.publishUpdateOne(detainee.centre));
+        .tap((detainee) => Centres.publishUpdateOne(detainee.centre));
     case Event.OPERATION_CHECK_IN:
     case Event.OPERATION_CHECK_OUT:
     case Event.OPERATION_REINSTATEMENT:
       return processEventDetainee(request_body)
-          .then((detainee) => Event.create(generateStandardEvent(detainee, request_body)))
-          .tap((event) => Centres.publishUpdateOne(event.centre));
+        .then((detainee) => Event.create(generateStandardEvent(detainee, request_body)))
+        .tap((event) => Centres.publishUpdateOne(event.centre));
+    case BedEvent.OPERATION_OUT_OF_COMMISSION:
+    case BedEvent.OPERATION_IN_COMMISSION:
+      return this.processBedEvent(request_body)
+        .tap((event) => Centres.publishUpdateOne(event.centre));
     default:
       throw new ValidationError('Unknown operation');
     }
   },
 
+  findAndPopulateCentre: event =>
+    Centres.getByName(event.centre).then((result) => _.merge(event, {centre_id: result.id})),
+
+  findAndPopulateDetainee: (event) =>
+    Detainee.getDetaineeByPersonIdAndCentre(event.single_occupancy_person_id, event.centre_id).then((result) => _.merge(event, {detainee: result})),
+
+  formatAndPopulateGender: (event) =>
+    _.merge(event, {gender: Bed.getGender(event.gender)}),
+
+  findOrCreateAndPopulateBed: (event) =>
+    Bed.findOrCreate({bed_ref: event.bed_ref, centre: event.centre_id}, {
+      centre: event.centre_id,
+      bed_ref: event.bed_ref,
+      gender: event.gender
+    }).then((bed) => _.merge(event, {bed_id: bed.id})),
+
+  findAndPopulateActiveStatus: (event) =>
+    BedEvent.findOne({bed: event.bed_id, timestamp: {'>=': event.timestamp}})
+      .then((result) =>
+        _.merge(event, {active: _.isUndefined(result)})),
+
+  reconcilePreviousBedEvents: (event) =>
+  event.active && BedEvent.deactivatePastBedEvents(event.bed_id, event.timestamp),
+
+  createBedEvent: (event) =>
+    BedEvent.create({
+      bed: event.bed_id,
+      active: event.active,
+      detainee: event.detainee,
+      timestamp: event.timestamp,
+      operation: event.operation,
+      reason: event.reason
+    }),
+
+  processBedEvent: function (event) {
```

Comment:
Sorry, this will work fine for linking the following sequence of events:

1. check in (person id = 1)
1. ooc (person id = 1)


it will however not produce the same result for:
 1. ooc (person id = 1)
 1. check in (person id = 1)